lightning/src/ln/channelmanager.rs

```diff
@@ -10911,43 +11073,22 @@ where
 					return None;
 				}
 
-				let error = match self.send_payment_for_verified_bolt12_invoice(
-					&invoice, payment_id,
-				) {
-					Err(Bolt12PaymentError::UnknownRequiredFeatures) => {
-						log_trace!(
-							logger, "Invoice requires unknown features: {:?}",
-							invoice.invoice_features()
-						);
-						InvoiceError::from(Bolt12SemanticError::UnknownRequiredFeatures)
-					},
-					Err(Bolt12PaymentError::SendingFailed(e)) => {
-						log_trace!(logger, "Failed paying invoice: {:?}", e);
-						InvoiceError::from_string(format!("{:?}", e))
-					},
-					Err(Bolt12PaymentError::UnexpectedInvoice)
-						| Err(Bolt12PaymentError::DuplicateInvoice)
-						| Ok(()) => return None,
-				};
-
-				match responder {
-					Some(responder) => Some((OffersMessage::InvoiceError(error), responder.respond())),
-					None => {
-						log_trace!(logger, "No reply path to send error: {:?}", error);
-						None
-					},
-				}
+				let res = self.send_payment_for_verified_bolt12_invoice(&invoice, payment_id);
+				handle_pay_invoice_res!(res, invoice, logger);
 			},
 			#[cfg(async_payments)]
-			OffersMessage::StaticInvoice(_invoice) => {
-				match responder {
-					Some(responder) => {
-						return Some((OffersMessage::InvoiceError(
-							InvoiceError::from_string("Static invoices not yet supported".to_string())
-						), responder.respond()));
+			OffersMessage::StaticInvoice(invoice) => {
```

Comment:
This is gonna take some extra work because `Event::InvoiceReceived` currently contains a `Bolt12Invoice` (i.e. not a static invoice), tracking this on #2298.